File: golang/dp_multidimensional/best_time_to_buy_and_sell_stock_transaction_fee.go
package main

func maxProfit(prices []int, fee int) int {
	n := len(prices)

	if n == 0 {
		return 0
	}

	hold, cash := -prices[0], 0

	for i := 1; i < n; i++ {
		hold = max(hold, cash-prices[i])
		cash = max(cash, hold+prices[i]-fee)
	}

	return cash
}

File: java/src/main/java/com/sanedge/interview/dp_multidimensional/BestTimeToBuyAndSellStockTransactionFee.java
package com.sanedge.interview.dp_multidimensional;

public class BestTimeToBuyAndSellStockTransactionFee {

    public int maxProfit(int[] prices, int fee) {
        int n = prices.length;

        if (n == 0) {
            return 0;
        }

        int hold = -prices[0];
        int cash = 0;

        for (int i = 1; i < n; i++) {
            hold = Math.max(hold, cash - prices[i]);
            cash = Math.max(cash, hold + prices[i] - fee);
        }

        return cash;
    }
}
